### PR TITLE
Routing

### LINKED ISSUE #15: use routes
- [x] Implement routes.ts file
- [x] Implement a router system that reflects URL to page element
- [x] Respond to navigation inside node list and push routes to history
- [x] Respond to click on breadcrumb and push routes to history

### PR DESCRIPTION
**Pull request by atahrijouti**

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,8 @@
+root = true
```
Comment (atahrijouti) [inline .editorconfig:1]: more on .editorconfig [on the highlighted portion on this page](https://blog.theodo.com/2019/08/why-you-should-use-eslint-prettier-and-editorconfig-together/#gatsby-focus-wrapper:~:text=.-,ESLint%2C%20Prettier%20and%20EditorConfig,EditorConfig%2C%20on%20the%20other%20hand%2C%20does%20not%20lint%20or%20format%20your%20code.%20It%20merely%20defines%20a%20standard%20code%20formatting%20style%20guide%20among%20all%20the%20IDEs%20and%20editors%20used%20within%20a%20team%20of%20developers.%20For%20instance%2C%20if%20a%20team%20uses%20two%20main%20editors%20such%20as%20Sublime%20Text%20and%20Visual%20Studio%20Code%2C%20EditorConfig%20allows%20them%20to%20define%20a%20common%20indentation%20pattern%20(spaces%20or%20tabs)%20within%20a%20single%20file.,-Customize%20the%20rules)

File: src/pages/app/app.css [hunk 2]
```diff
@@ -12,7 +12,10 @@ a:hover {
   text-decoration: underline;
 }
 
-html, body, #root, .app  {
+html,
```
Comment (atahrijouti) [inline src/pages/app/app.css:15]: just prettier formatting

File: src/pages/app/explorer/node/node-icon.ts [hunk 3]
```diff
@@ -8,22 +8,28 @@ import { Node, NodeType, state } from "~pages/app/state"
 
 export function nodeIcon(node: Node): string {
   switch (node.type) {
-    case NodeType.FILE: return getFileImage(node)
-    case NodeType.FOLDER: return getFolderImage(node)
-    default: return unknownImage
+    case NodeType.FILE:
```
Comment (atahrijouti) [inline src/pages/app/explorer/node/node-icon.ts:11]: just prettier formating

File: src/pages/app/navigation-bar/control-button.ts [hunk 4]
```diff
@@ -1,14 +1,23 @@
...
+import cx from "classnames"
 
 type ControlButtonProps = {
-  label: string
+  label: Node | string
   image: string
-}
-export const ControlButton = ({ label, image }: ControlButtonProps) => {
+  className?: string
+} & Partial<HTMLButtonElement>
```
Comment (atahrijouti) [inline src/pages/app/navigation-bar/control-button.ts:8]: making button behave as a `HTMLButtonElement` and expect the same attributes from it

File: src/pages/app/navigation-bar/control-button.ts [hunk 5]
```diff
@@ -1,14 +1,23 @@
...
+  label,
+  image,
+  className,
+  type = "button",
+  ...props
+}: ControlButtonProps) => {
   return h(
     "button",
-    { className: "control-button" },
+    { className: cx("control-button", className), ...props },
```
Comment (atahrijouti) [inline src/pages/app/navigation-bar/control-button.ts:19]: accept a custom class, spill all props passed as arguments to attributes of the html button

File: src/pages/app/navigation-bar/index.ts [hunk 6]
```diff
@@ -122,7 +130,35 @@ export function NavigationBar() {
       newFolderBtn,
       newFileBtn,
       renameNodeBtn,
-      deleteNodeBtn
+      deleteNodeBtn,
+      ControlButton({
```
Comment (atahrijouti) [inline src/pages/app/navigation-bar/index.ts:134]: few buttons for testing routing with gui

File: src/pages/app/navigation-bar/navigation-bar.scss [hunk 7]
```diff
@@ -24,10 +24,10 @@
   font-size: 13px;
   border: 1px solid transparent;
   &:hover {
-    background: #E5F3FF;
-    border: 1px solid #CCE8FF;
+    background: #e5f3ff;
```
Comment (atahrijouti) [inline src/pages/app/navigation-bar/navigation-bar.scss:27]: just prettier reformatting

File: src/router/index.ts [hunk 8]
```diff
@@ -1,17 +1,101 @@
...
+    return path
+  }
+  return path.replace(/\/$/, "")
+}
+
+function getPathFromWindowUrl() {
+  return cleanPath(window.location.pathname)
+}
+
+function matchPath(path: string) {
```
Comment (atahrijouti) [inline src/router/index.ts:55]: goes through known routes defined in `ROUTES` configuration, and tries to match the path given it with an existing route

File: src/router/index.ts [hunk 9]
```diff
@@ -1,17 +1,101 @@
...
+  return ROUTES.find((route) => path.match(route.pattern))
+}
+
+let currentRouteHandler: (() => HTMLElement) | null = null
+let currentRoutePage: HTMLElement | null = null
```
Comment (atahrijouti) [inline src/router/index.ts:60]: `currentRouteHandler` is the function to execute when a match was found for the current route. `currentRoutePage` is what comes out of calling `currentRouteHandler`.  `currentRoutePage`  is then called if the next time we hit a route that would call the same `currentRouteHandler`. In this case, we simply dispatch a `AppEvent.ROUTE_CHANGED` event, and not re-render the `currentRoutePage` completely

File: src/router/index.ts [hunk 10]
```diff
@@ -1,17 +1,101 @@
...
+import { rootElement } from "~index"
 
 export type LinkProps = {
   path: string
   title: string
   className?: string
+  children: Node | Node[]
 }
-export function Link({ path, title, className }: LinkProps) {
+export function Link({ path, title, children, className }: LinkProps) {
```
Comment (AmineTir) [inline src/router/index.ts:13]: What does this Link() method & where is it used?

File: src/router/index.ts [hunk 11]
```diff
@@ -1,17 +1,101 @@
...
   function handlClick(e: MouseEvent) {
```
Comment (AmineTir) [inline src/router/index.ts]: typo

File: src/router/index.ts (hunk 10 above)
Comment (atahrijouti) [reply src/router/index.ts:13]: It's not currently used, but from my experience there is usually need for an `<a />` that triggers the router

File: src/router/index.ts (hunk 11 above)
Comment (atahrijouti) [inline src/router/index.ts]:
```suggestion
  function handleClick(e: MouseEvent) {
```

File: src/router/index.ts [hunk 12]
```diff
@@ -1,17 +1,101 @@
...
     e.preventDefault()
-    history.pushState(null, title, path)
+    navigateTo(path, title)
   }
 
-  return h("a", { className: cx("link", className), href: path, ["onclick"]: handlClick, title })
+  return h(
+    "a",
+    { className: cx("link", className), href: path, ["onclick"]: handlClick, title },
```
Comment (atahrijouti) [inline src/router/index.ts]:
```suggestion
    { className: cx("link", className), href: path, ["onclick"]: handleClick, title },
```

**Review by AmineTir — APPROVED** [review]

**Pull request merged**